### ISSUE TITLE
[BUG] account: Concurrent update creating more than one invoice/payment - Odoo>=14.0

### ISSUE THREAD

**Issue by moylop260** [odoo/odoo#90465]
# Impacted versions:
14.0, 15.0 and current master (After the following commit: https://github.com/odoo/odoo/commit/dfd01b8c5c7e1177f37bf199790a0732a61eed78)

# Steps to reproduce:
Install the following modules:
 - `-i sale_subscription,website_sale,sale_management,website_sale_management,payment_test`
Configure product:
 - `Office Cleaning Subscription (Monthly)` to `publish in website`
Configure in the settings sale to Automatic Invoice
 - Invoicing policy: Ordered
Apply then choose
 - Automatic Invoice: True
Apply

Make a bot to make online sales

Check the following javascript:
 - https://gist.github.com/moylop260/7e39a03f119d4be148385b40c3a4bf76

And run the following command:
 - `k6 run --vus 1 --iterations 30 checkout.js --duration=120m`

Then update all the subscriptions generated to set an past date.
```python
env['sale.subscription'].search([('stage_id', '=', 2)]).write({
  'recurring_next_date': '2022-05-01',
  'tag_ids': False,
})
```

Enable the following cron using a close date and active:
 - `Sale Subscription: generate recurring invoices and payments`

Run the bot again at the same time than cron is running:
 - `k6 run --vus 1 --iterations 30 checkout.js --duration=120m`

# Current behavior:

<details>
  <summary>See the following `concurrent update` error in the logs:</summary>

```txt
2022-05-04 00:34:25,976 23 ERROR 15142441-14-0-all odoo.sql_db: bad query: 
            UPDATE account_move SET write_date = write_date WHERE id = (
                SELECT id FROM account_move
                WHERE journal_id = 2 AND name != '/' AND date_trunc('month', date::timestamp without time zone) = date_trunc('month', '2022-05-04'::date)  AND move_type NOT IN ('out_refund', 'in_refund')  AND id != 319 
                AND sequence_prefix = (SELECT sequence_prefix FROM account_move WHERE journal_id = 2 AND name != '/' AND date_trunc('month', date::timestamp without time zone) = date_trunc('month', '2022-05-04'::date)  AND move_type NOT IN ('out_refund', 'in_refund')  AND id != 319  ORDER BY id DESC LIMIT 1)
                ORDER BY sequence_number DESC
                LIMIT 1
            )
            RETURNING name;
        
ERROR: could not serialize access due to concurrent update
 
2022-05-04 00:34:25,990 23 ERROR 15142441-14-0-all odoo.addons.sale_subscription.models.sale_subscription: Traceback (most recent call last):
  File "/data/build/enterprise/sale_subscription/models/sale_subscription.py", line 963, in _recurring_create_invoice
    new_invoice = Invoice.create(invoice_values)
  File "<decorator-gen-329>", line 2, in create
  File "/data/build/odoo/odoo/api.py", line 347, in _model_create_multi
    return create(self, [arg])
  File "/data/build/odoo/addons/purchase/models/account_invoice.py", line 102, in create
    moves = super(AccountMove, self).create(vals_list)
  File "<decorator-gen-240>", line 2, in create
  File "/data/build/odoo/odoo/api.py", line 348, in _model_create_multi
    return create(self, arg)
  File "/data/build/odoo/addons/account/models/account_move.py", line 1980, in create
    rslt = super(AccountMove, self).create(vals_list)
  File "<decorator-gen-135>", line 2, in create
  File "/data/build/odoo/odoo/api.py", line 348, in _model_create_multi
    return create(self, arg)
  File "/data/build/odoo/addons/mail/models/mail_thread.py", line 264, in create
    threads = super(MailThread, self).create(vals_list)
  File "<decorator-gen-64>", line 2, in create
  File "/data/build/odoo/odoo/api.py", line 348, in _model_create_multi
    return create(self, arg)
  File "/data/build/odoo/odoo/addons/base/models/ir_fields.py", line 534, in create
    recs = super().create(vals_list)
  File "<decorator-gen-13>", line 2, in create
  File "/data/build/odoo/odoo/api.py", line 348, in _model_create_multi
    return create(self, arg)
  File "/data/build/odoo/odoo/models.py", line 3876, in create
    records = self._create(data_list)
  File "/data/build/odoo/odoo/models.py", line 4034, in _create
    field.create([
  File "/data/build/odoo/odoo/fields.py", line 3043, in create
    self.write_batch(record_values, True)
  File "/data/build/odoo/odoo/fields.py", line 3069, in write_batch
    return self.write_real(records_commands_list, create)
  File "/data/build/odoo/odoo/fields.py", line 3245, in write_real
    flush()
  File "/data/build/odoo/odoo/fields.py", line 3207, in flush
    comodel.create(to_create)
  File "<decorator-gen-316>", line 2, in create
  File "/data/build/odoo/odoo/api.py", line 348, in _model_create_multi
    return create(self, arg)
  File "/data/build/odoo/addons/membership/models/account_move.py", line 84, in create
    lines = super(AccountMoveLine, self).create(vals_list)
  File "<decorator-gen-241>", line 2, in create
  File "/data/build/odoo/odoo/api.py", line 348, in _model_create_multi
    return create(self, arg)
  File "/data/build/odoo/addons/account/models/account_move.py", line 4081, in create
    moves._check_balanced()
  File "/data/build/odoo/addons/account/models/account_move.py", line 1798, in _check_balanced
    self.env['account.move.line'].flush(self.env['account.move.line']._fields)
  File "/data/build/odoo/odoo/models.py", line 5462, in flush
    self.recompute(fnames, records=records)
  File "/data/build/odoo/odoo/models.py", line 5928, in recompute
    process(field)
  File "/data/build/odoo/odoo/models.py", line 5899, in process
    field.recompute(recs)
  File "/data/build/odoo/odoo/fields.py", line 1155, in recompute
    self.compute_value(recs)
  File "/data/build/odoo/odoo/fields.py", line 1177, in compute_value
    records._compute_field_value(self)
  File "/data/build/odoo/odoo/models.py", line 4071, in _compute_field_value
    field.compute(self)
  File "/data/build/odoo/odoo/fields.py", line 555, in _compute_related
    record[self.name] = self._process_related(value[self.related_field.name])
  File "/data/build/odoo/odoo/models.py", line 5686, in __getitem__
    return self._fields[key].__get__(self, type(self))
  File "/data/build/odoo/odoo/fields.py", line 969, in __get__
    self.recompute(record)
  File "/data/build/odoo/odoo/fields.py", line 1155, in recompute
    self.compute_value(recs)
  File "/data/build/odoo/odoo/fields.py", line 1177, in compute_value
    records._compute_field_value(self)
  File "/data/build/odoo/addons/mail/models/mail_thread.py", line 410, in _compute_field_value
    return super()._compute_field_value(field)
  File "/data/build/odoo/odoo/models.py", line 4069, in _compute_field_value
    getattr(self, field.compute)()
  File "/data/build/odoo/addons/account/models/account_move.py", line 1158, in _compute_name
    highest_name = self[0]._get_last_sequence() if self else False
  File "/data/build/odoo/addons/account/models/sequence_mixin.py", line 179, in _get_last_sequence
    self.env.cr.execute(query, param)
  File "<decorator-gen-3>", line 2, in execute
  File "/data/build/odoo/odoo/sql_db.py", line 101, in check
    return f(self, *args, **kwargs)
  File "/data/build/odoo/odoo/sql_db.py", line 298, in execute
    res = self._obj.execute(query, params)
psycopg2.errors.SerializationFailure: could not serialize access due to concurrent update

 
2022-05-04 00:34:25,997 23 ERROR 15142441-14-0-all odoo.addons.sale_subscription.models.sale_subscription: Error during renewal of subscription SUB004 (No payment recorded.) 
```

</details>

# Expected behavior:

It is expected since that the `account.move` records (payment, invoices) needs to lock the last record to get a new number

It is similar to use no-gap ir.sequence configuration

It was easy to fix for Odoo <=13.0 changing the ir.sequence to Standard since that no-gap is not a good idea for instances with many concurrent users creating payments and invoices

And for a few countries the gap are not a big deal than errors in online checkout users or subscriptions raising errors because of the number of invoice/payment

In fact, for MX the XML electronic invoice UUID is the official reference/number for the government the account_move.name is only a reference where the GAPs are allowed

New projects already migrated to >=14.0 we have a lot of issues related with this since that the subscription are in a invalid state, the front-end user are raising errors after making the payment and then try to make the payment again and again then they have double charges since that the original one was reverted because of concurrent update.

We even have deadlocks for automatic process creating invoice->payment or payment->invoice

So, It is a catastrophic specially for customers where they have too many users creating invoices/payments

The expected behavior is a option to allow gaps (standard ir.sequence) way

From performance perspective, we are creating Odoo even faster for each version 
but after >=13.0 is only allowing to create one invoice/payment each time
so it is a big deal for performance since that if you can support 5k concurrent user doing checkout process you will have a limit to create the invoices to only one by one or concurrency errors

So, we have only one door with size of 1 person but 5k are in the queue waiting for release the account.move

So, the all the workers are used rapidly waiting for releases so the system is down since that there are not more workers to attend new requests (You can use `--vus 3` in the k6 command to see how the runbot is not responding anymore after that, notice only 3 users even if they are using 3 workers in runbot)

It is a crazy idea.

Now, try to explain to your customers recently migrated to improve the performance of theirs system... so they don't believe us, so I have created this issue in order to see the options and have a opportunity to believe us

# Video/Screenshot link (optional):
See the following video using runbot to raise the concurrent errors
 - [![IMAGE ALT TEXT HERE](https://img.youtube.com/vi/kDVn8FPs6Us/0.jpg)](https://www.youtube.com/watch?v=kDVn8FPs6Us)

**Comment by jco-odoo** [comment]
@moylop260 If we want to solve this problem, we need to do it for all countries, because most countries need the no-gap (MX is rather an exception here) and in terms of locking, it is the same thing as happened with the sequences before.  Already, posting an invoice should not take much time and maybe there is a solution by locking as late as possible.  (so the locking time is as short as possible)

**Comment by jco-odoo** [comment]
(+ we might still be able to inherit the methods to avoid the locking as well)

**Comment by jco-odoo** [comment]
You have it for the numbering in a lot of other latam countries.  Belgium/FR have it too.  (or it is at least preferable for auditing) And what about minimizing the locked time?  If you have only 10 ms of locked time, you should be able to do quite some invoices / second.  @william-andre

**Comment by lem8r** [comment]
In our case most Odoo projects are not using accounting for legal purposes. However most of them are  based on OE.

There are few projects that are using Odoo to import and reconcile many invoices and bank statements.
To speed up invoices import process we are running many crons. 
Bottle necks start to arise. Gapless sequence, increment of customer rank. Slow reconcile sql query that is too smart. Accounting dashboard is slow if there are hundred of thousands open invoices and statement lines. 

All of that we fixed with custom code but it was hard to explain to the customer why ERP system is not designed for large companies out of the box.

Luckily increment of customer rank is fixed now.

**Comment by moylop260** [comment]
FYI after migrating to v14.0 the postgresql.log raises the following error:

```
ERROR:  40P01: deadlock detected
DETAIL:  Process 1875429 waits for ShareLock on transaction 5750251; blocked by process 1875141.
	Process 1875141 waits for ShareLock on transaction 5750249; blocked by process 1875429.
	Process 1875429: 
	            UPDATE account_move SET write_date = write_date WHERE id = (
	                SELECT id FROM account_move
	                WHERE journal_id = 104 AND name != '/' AND date_trunc('year', date::timestamp without time zone) = date_trunc('year', '2022-05-03'::date)  AND sequence_prefix !~ '^(?:.*?)(?:((?<=\D)|(?<=^))((20|21)\d{2}|(\d{2}(?=\D))))(?:\D*?)(?:(0[1-9]|1[0-2]))(?:\D+?)$'  AND id != 2615179 
	                AND sequence_prefix = (SELECT sequence_prefix FROM account_move WHERE journal_id = 104 AND name != '/' AND date_trunc('year', date::timestamp without time zone) = date_trunc('year', '2022-05-03'::date)  AND sequence_prefix !~ '^(?:.*?)(?:((?<=\D)|(?<=^))((20|21)\d{2}|(\d{2}(?=\D))))(?:\D*?)(?:(0[1-9]|1[0-2]))(?:\D+?)$'  AND id != 2615179  ORDER BY id DESC LIMIT 1)
	                ORDER BY sequence_number DESC
	                LIMIT 1
	            )
	            RETURNING name;
	        
	Process 1875141: 
	            UPDATE account_move SET write_date = write_date WHERE id = (
	                SELECT id FROM account_move
	                WHERE journal_id = 1 AND name != '/' AND date_trunc('month', date::timestamp without time zone) = date_trunc('month', '2022-05-03'::date)  AND id != 2615176 
	                AND sequence_prefix = (SELECT sequence_prefix FROM account_move WHERE journal_id = 1 AND name != '/' AND date_trunc('month', date::timestamp without time zone) = date_trunc('month', '2022-05-03'::date)  AND id != 2615176  ORDER BY id DESC LIMIT 1)
	                ORDER BY sequence_number DESC
	                LIMIT 1
	            )
	            RETURNING name;
```

Notice that the journal 1 is type `sale` (account.move related to invoice)
and the journal 104 is type `bank` (account.move related to payment)

It is related with last comments related with payment deadlock too

**Comment by william-andre** [comment]
Please see https://github.com/odoo/odoo/pull/91260

If you could write a test, preferably launchable from the testing suite, like `odoo.addons.account.tests.test_sequence_mixin.test_sequence_concurency`, or using XMLRPC (and easy to run in local in all versions without having to hardcode values depending on demo data etc.), that would be great.
The script you provided is just a pain to run because
* demo data
* need manual setup
* need external tooling

I tried to run it but there were too many things to tweek.

**Comment by moylop260** [comment]
@william-andre 

I have created the following PR:
 - https://github.com/odoo/odoo/pull/91525

In order to raises the current concurrent issues for v14.0

It is related with account.move only

But the issues are even more when the raise comes from another model since that the records related could be inconsistent (subscription, payment transaction, sale order, pos order and so on)

**Comment by moylop260** [comment]
Since that it issue was related to use standard or not gaps the new one is related to all the new concurrency errors even using no-gaps for v13.0 and v14.0

Superseded by https://github.com/odoo/odoo/issues/91873
